common_reference broken with MSVC 19.22

The latest MSVC update (v19.22) breaks our implementation of `common_reference`, to the extent that just `#include <nanorange.hpp>` generates

    error C2064: term does not evaluate to a function taking 0 arguments
    note: see reference to alias template instantiation 'nano::ranges::detail::cond_res_t<copy_cv<X,Y>::type&,copy_cv<Y,X>::type&>' being compiled

(See https://godbolt.org/z/VXv5gY)

The offending line seems to be https://github.com/tcbrindle/NanoRange/blob/master/include/nanorange/detail/common_reference.hpp#L68. I'm not sure of the cause -- this should be in a SFINAE context, so if evaluation of the alias causes a substitution failure, then the partial specialisation should just be ignored... or at least that's the idea.

It works correctly with Clang, GCC and previous versions of MSVC, so it may just be a compiler bug, or it may be that MSVC is now correctly diagnosing a problem that we previously got away with. Either way, this prevents NanoRange being used with the latest Windows compiler, so it needs to be fixed ASAP.

--- a/include/nanorange/detail/common_reference.hpp
+++ b/include/nanorange/detail/common_reference.hpp
@@ -43,9 +43,28 @@ using copy_cv_t = typename copy_cv<T, U>::type;
 template <typename T>
 using cref_t = std::add_lvalue_reference_t<const std::remove_reference_t<T>>;
 
+// Workaround for "term does not evaluate to a function taking 0 arguments"
+// error in MSVC 19.22 (issue #75)
+#if defined(_MSC_VER) && _MSC_VER >= 1922
+template <typename, typename, typename = void>
+struct cond_res {};
+
+template <typename T, typename U>
+struct cond_res<T, U, std::void_t<decltype(false ? std::declval<T (&)()>()()
+                                                 : std::declval<U (&)()>()())>>
+{
+    using type = decltype(false ? std::declval<T (&)()>()()
+                                : std::declval<U (&)()>()());
+};
+
+template <typename T, typename U>
+using cond_res_t = typename cond_res<T, U>::type;
+#else
 template <typename T, typename U>
 using cond_res_t = decltype(false ? std::declval<T (&)()>()()
                                   : std::declval<U (&)()>()());
+#endif
+
 
 // For some value of "simple"
 template <typename A, typename B,

